controller/src/main/java/io/pravega/controller/eventProcessor/impl/EventProcessorGroupImpl.java

```diff
@@ -204,8 +199,51 @@ protected void shutDown() {
         synchronized (lock) {
             long traceId = LoggerHelpers.traceEnterWithContext(log, this.objectId, "shutDown");
             try {
-                cleanUpCheckpointStore();
+                // If any of the following operations error out, the fact is just logged.
+                // And some other controller process would clean-up the reader group data
+                // its readers and their position objects from checkpoint store.
+
+                // Seal the Reader Group
+                Map<String, Position> readerPositions = ImmutableMap.of();
+                try {
+                    readerPositions = sealReaderGroup(actorSystem.getProcess());
+                } catch (CheckpointStoreException e) {
+                    log.warn("Error sealing Reader Group {}.", readerGroup.getGroupName(), e);
+                }
+
+                // Initiate stop on all event processor cells
+                for (EventProcessorCell<T> cell : eventProcessorMap.values()) {
+                    log.info("Terminating event processor cell: {}", cell);
+                    cell.stopAsync();
+                }
+                // Await termination of cells
+                for (EventProcessorCell<T> cell : eventProcessorMap.values()) {
+                    try {
+                        cell.awaitTerminated();
+                        log.info("Termination of event processor cell: {} completed successfully.", cell);
+                    } catch (Exception e) {
+                        log.warn("Failed terminating event processor cell {}.", cell, e);
+                        throw e;
```

Comment:
If an exception is thrown here then the controller would not remove the checkpointStore reader? The older code would attempt to remove the reader(s) fromthe checkpointStore even in such a scenario. Is this change in behaviour intended?

Reply:
The addition of line `throw e`  here is intentional. If we fail to stop the EventProcessor cell there is a possibility that the Readers in corresponding reader group were not made offline and without that we do not want t remove them from the checkpoint store.
The expectation here is that another controller will sweep entries for these readers and declare them offline as part of notifyProcessFailure.......
Alternatively, if invoking the readerOffline() is an idempotent operation we could attempt doing it again here by invoking cleanupReaderGroup...